### ISSUE TITLE
load_setuptools() in meta.yaml spawns multiple processes and confuses everything

### ISSUE THREAD

**Issue by jankatins** [conda/conda-build#1061]
This is part of https://github.com/matplotlib/matplotlib/pull/6520 which after fixing the ÀPI change of `CONDA_DEFAULT_ENV` now can't find `cl.exe` during the conda package build (it's totaly fine during the normal build, not sure what condabuild is doing to "forget" that it is findable...).

See this appveyor build log:
https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o

**Comment by msarahan** [comment]
Weird, I thought https://github.com/conda/conda-build/blob/1.21.x/tests/test_build_recipes.py#L206-L223 and https://github.com/conda/conda-build/blob/1.21.x/tests/test-recipes/metadata/_build_msvc_compiler/bld.bat would catch that.  In those tests, msvc_compiler is being set in the meta.yaml file, thus defining the MSSdk and DISTUTILS_USE_SDK variables.  Is distutils possibly not respecting the compiler that's available because these are not set?

**Comment by msarahan** [comment]
It may be that we just want to always set those, and bypass any logic that setuptools/distutils/whatever has.

**Comment by msarahan** [comment]
In looking at this in more detail, it seems like more than one build is occurring simultaneously:

* https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L2945
* https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3066
* https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3182
* https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3296

What's going on there?

It does appear that the VS 2010 stuff is in PATH: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3429

I don't know what any of this VCTools stuff is from, but it also looks repeated: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3693

Still have VS 2010 in PATH: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3774

This should do nothing: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3861

Unless this is breaking things, I think Setuptools is my best guess: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1878/job/vmo3gr4po2wekj3o#L3863

Can you please output the value of PATH as the first command in your bld.bat?

**Comment by janschulz** [comment]
My guess fot the "funny output" was that I use `{% set data = load_setuptools() %}` and this before triggered lock errors as setup.py uses multiprocessing is used in "configure" and it seems it recreates the environment which in this case includes conda-build internals.

**Comment by janschulz** [comment]
I've also added a `set` as the first command in bld.bat: https://github.com/matplotlib/matplotlib/pull/6520/commits/dcbb6f89d5790a9a927004bffbd03f18cc144131 and the appveyor runs: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1885

**Comment by msarahan** [comment]
well, it's not in PATH: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1885/job/eqvf8l4r0xqs3v3r#L3967

Here's the last place I see it in PATH: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1885/job/eqvf8l4r0xqs3v3r#L3766

That points primarily to the activate script as the culprit, but I don't think that's true: its stored backup PATH does not have VS 2010 in it: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1885/job/eqvf8l4r0xqs3v3r#L3911

Sorry, I really don't think we can reason about this further until you have straightened out the multiple builds.  It is too hard to know what we can trust, and what is meaningless noise.

**Comment by msarahan** [comment]
We have figured this out, we think.  If you have an environment active when starting a build, then the conda activate scripts discard that environment prior to activating a new one.  When VS is active in the former, it no longer works in the latter.

We'll try to fix this quickly.

**Comment by msarahan** [comment]
This PR should hopefully address the issue.  Feedback welcome. https://github.com/conda/conda/pull/2917

**Comment by janschulz** [comment]
I'm currently doing a rebuild in the matplotlib build... let's see :-)

**Comment by msarahan** [comment]
I don't think that pr is in a tagged release yet. Not sure if this is going to be any different.

**Comment by janschulz** [comment]
oups, ok.

Anyway, I also saw that I'm currently doing a  `deactivate` before starting the conda-build process:

https://github.com/janschulz/matplotlib/blob/dcbb6f89d5790a9a927004bffbd03f18cc144131/appveyor.yml#L152-L156

**Comment by msarahan** [comment]
Yeah,  that doesn't really reaffirm my theory. If the pr does not fix the problem, then I'm afraid I'm back to my request that we get the duplication strangeness sorted out before debugging further.

**Comment by janschulz** [comment]
Ok, I removed the `load_setuptools()` call, lets see if this fixes things... https://github.com/matplotlib/matplotlib/pull/6520 https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1909

**Comment by janschulz** [comment]
Now the only test failure is this:

```
C:\Miniconda-x64\conda-bld\work>call "C:\Program Files (x86)\Microsoft Visual Studio 9.0\VC\vcvarsall.bat" amd64 
Setting environment for using Microsoft Visual Studio 2008 x64 tools.
The input line is too long.
The syntax of the command is incorrect.
Command failed: C:\windows\system32\cmd.exe /c bld.bat
Command exited with code 1
```

[Only on py27, 64bit]

https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1912/job/88gjt5vm2mbymedj

**Comment by janschulz** [comment]
I removed teh conda-build command from the original PR as a workaround. The failing appveyor build is now in https://github.com/matplotlib/matplotlib/pull/6682

**Comment by msarahan** [comment]
What's a good way for me to debug this? Fork the branch and try locally, then on Appveyor?  The CMD.exe limit for any given command is ~8000 characters.  I have no idea how we would be getting anywhere close to that.  The longest PATH entries I have ever seen were ~1500 characters.

**Comment by janschulz** [comment]
This PATH has already 3400 chars :-/ But what is more surprising that it only happens now, not with the earlier builds :-/

I've pushed a commit which resets PATH to a shorter one... (and one which pushes the 64bit py27 to the front and disables the normal tests, so that the runtime is faster)

Not sure which one is faster:
-> https://ci.appveyor.com/project/JanSchulz/matplotlib/build/1.0.195
-> https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1973

Note that it uses the real VS, not the python tools, which I "fix" here: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1958/job/53c53pg0mtqpoacw#L145

**Comment by msarahan** [comment]
Wait, how are you using "the real VS"?  AppVeyor doesn't have it installed
for 64-bit compiling with VS 2008.  They only have the express version of
VS 2008: https://www.appveyor.com/docs/installed-software#visual-studio-2008

What are you hoping to accomplish with copying that batch file?  You might
be helping something find the VC Compiler for Python, but I don't think
you're using the "real VS."

Even 3400 characters should not be a problem based on what I've read, and
what @mingwandroid has experimented with.  I'm also confused on why the
other builds succeed: are their paths shorter?

**Comment by janschulz** [comment]
> Wait, how are you using "the real VS"?  AppVeyor doesn't have it installed
for 64-bit compiling with VS 2008.  They only have the express version of
VS 2008: https://www.appveyor.com/docs/installed-software#visual-studio-2008

Yep, that's why I copied the file: [code](https://github.com/janschulz/matplotlib/blob/523e32cd3321bd3aadd711156c4083a5cc037de9/appveyor.yml#L75-L80). That's at least the workaround I had in place here on my own windows install and it worked before when conda build wasn't able to recognize the python tools.

**Comment by msarahan** [comment]
PS, I'm adding https://ci.appveyor.com/project/msarahan/matplotlib/build/1.0.2 - trying to get an artifact of the bld.bat.  I'm not certain whether this error is in our activate scripts or somewhere in the VS activation scripts.  Last output is from them, but that doesn't mean much.

**Comment by msarahan** [comment]
Wow, is conda failing this badly: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1974/job/jpmhkbbx6adn60ej

**Comment by janschulz** [comment]
Looks more like an anaconda.org / conda-forge problem?

**Comment by msarahan** [comment]
https://github.com/conda/conda/issues/3004

**Comment by janschulz** [comment]
re conda error: outch, you really need more regression/release tests...

This build (same commit, other account) still worked: https://ci.appveyor.com/project/JanSchulz/matplotlib/build/job/485opc5usdp686eq

-> Seems that removing the workaround for py27/64bit VS made it work. I still don't get why that happened now but now a few days/weeks before...?

> Deprecation warning: support for __conda_version__ will be removed in Conda build 2.0.Try Jinja templates instead: http://conda.pydata.org/docs/building/environment-vars.html#git-environment-variables

Outch, that will be a problem for matplotlib if the ci tools should keep the automatic versions :-( Any chance for reviving https://github.com/conda/conda-build/issues/714 ?

**Comment by msarahan** [comment]
Regarding #714 - yes.  If I want to get rid of post-build stuff, I need to provide people with every possible way (within reason) of getting the version info before the build - from source or otherwise.

Please start a separate issue for that, but yes, let's work together on what a solution looks like for you, and what assumptions/restrictions there are.

**Comment by msarahan** [comment]
> Seems that removing the workaround for py27/64bit VS made it work. I still don't get why that happened now but now a few days/weeks before...?

some of what has changed recently has been my fumbling with ```call``` in scripts to make the activate scripts work and not exit the build.  I ended up finding that using ```call``` for every bat file within bld.bat was best.  bld.bat itself is not called with call - just ```cmd.exe /c bld.bat```.  I wonder if this changed the error behavior of your workaround.

Please keep me posted on whether any other problems arise.

**Comment by msarahan** [comment]
I'm going to close this issue, since removing your workaround worked.  I'll reopen it if you find otherwise later.

**Comment by janschulz** [comment]
Note that the origonal problem is still not gone: you can't use `load_setuptools()` with the current setup.py in matplotlib (see above) and the curren workaround (building a `__conda_version__` file) is going to be removed. So this bug will come up when you remove that feature...

So for that it's IMO either reopening this issue (=fix using matplotlibs setup.py with `load_setuptools()`) or #714 (give mpl a way to use versioneer directly to get a version)

**Comment by msarahan** [comment]
I think the original issue was not finding cl.exe. I  think that is fixed.

I am open to both reopening #714 and opening a new issue for misbehavior of
load_setuptools

**Comment by janschulz** [comment]
The "cl.exe not found" is triggered by the `load_setuptools()` problem as far as I understand it: load_setuptools in this cases spawns multiple subprocesses which all try to recreate the conda build internals which then somehow interfere with each other so that the build doesn't find cl.exe. 

In earlier builds (=older `conda build`) that "only" ended in locking errors as the "new" ones were running into locks which made the subprocess error out due to timeouts but mpl treats the timeouts as "not available" and let `setup()` return and so load_setuptools()` got the version.

**Comment by msarahan** [comment]
OK,  thanks. I am reopening and renaming this issue accordingly.
